# 📝 Exercise M1.03

The goal of this exercise is to compare the performance of our classifier in
the previous notebook (roughly 81% accuracy with `LogisticRegression`) to
some simple baseline classifiers. The simplest baseline classifier is one
that always predicts the same class, irrespective of the input data.

- What would be the score of a model that always predicts `' >50K'`?
- What would be the score of a model that always predicts `' <=50K'`?
- Is 81% or 82% accuracy a good score for this problem?

Use a `DummyClassifier` and do a train-test split to evaluate
its accuracy on the test set. This
[link](https://scikit-learn.org/stable/modules/model_evaluation.html#dummy-estimators)
shows a few examples of how to evaluate the generalization performance of these
baseline models.

In [1]:
import pandas as pd

adult_census = pd.read_csv("../datasets/adult-census.csv")

We will first split our dataset to have the target separated from the data
used to train our predictive model.

In [2]:
target_name = "class"
target = adult_census[target_name]
data = adult_census.drop(columns=target_name)

We start by selecting only the numerical columns as seen in the previous
notebook.

In [3]:
numerical_columns = [
    "age", "capital-gain", "capital-loss", "hours-per-week"]

data_numeric = data[numerical_columns]

Split the data and target into a train and test set.

In [4]:
from sklearn.model_selection import train_test_split

# Write your code here.
data_train, data_test, target_train, target_test = train_test_split(data_numeric, target, random_state=42, test_size=0.25)

In [5]:
print(f"Number of samples in testing: {data_test.shape[0]} => "
      f"{data_test.shape[0] / data_numeric.shape[0] * 100:.1f}% of the"
      f" original set")

Number of samples in testing: 12211 => 25.0% of the original set


In [6]:
print(f"Number of samples in training: {data_train.shape[0]} => "
      f"{data_train.shape[0] / data_numeric.shape[0] * 100:.1f}% of the"
      f" original set")

Number of samples in training: 36631 => 75.0% of the original set


Use a `DummyClassifier` such that the resulting classifier will always
predict the class `' >50K'`. What is the accuracy score on the test set?
Repeat the experiment by always predicting the class `' <=50K'`.

Hint: you can set the `strategy` parameter of the `DummyClassifier` to
achieve the desired behavior.

In [15]:
from sklearn import set_config
set_config(display='diagram')

In [20]:
from sklearn.dummy import DummyClassifier

# Write your code here.
baseline_model = DummyClassifier(strategy = 'constant', constant = ' >50K', random_state=0)

In [21]:
baseline_model.fit(data_train, target_train)

DummyClassifier(constant=' >50K', random_state=0, strategy='constant')

In [24]:
accuracy = baseline_model.score(data_test, target_test)
print(f"Accuracy of Dummy Classifier for >50K: {accuracy:.3f}")

Accuracy of Dummy Classifier for >50K: 0.234


In [25]:
baseline_model_2 = DummyClassifier(strategy = 'constant', constant = ' <=50K', random_state=0)
baseline_model_2.fit(data_train, target_train)
accuracy = baseline_model_2.score(data_test, target_test)
print(f"Accuracy of Dummy Classifier for <=50K: {accuracy:.3f}")

Accuracy of Dummy Classifier for <=50K: 0.766


We observe that this model has an accuracy higher than 0.5. This is due to the fact that we have 3/4 of the target belonging to low-revenue class.

Therefore, any predictive model giving results below this dummy classifier will not be helpful.

In [26]:
adult_census["class"].value_counts()

 <=50K    37155
 >50K     11687
Name: class, dtype: int64

In [27]:
(target == " <=50K").mean()

0.7607182343065395

So the LogisticRegression accuracy (roughly 81%) seems better than the DummyClassifier accuracy (roughly 76%). In a way it is a bit reassuring, using a machine learning model gives you a better performance than always predicting the majority class, i.e. the low income class " <=50K".